test/spoof/spoof.go

```diff
@@ -174,3 +189,33 @@ func (sc *SpoofingClient) Poll(req *http.Request, inState ResponseChecker) (*Res
 
 	return resp, err
 }
+
+// LogZipkinTrace provides support to log Zipkin Trace for param: traceID
+func (sc *SpoofingClient) LogZipkinTrace(traceID string) error {
+
+	if zipkin.CheckZipkinPortAvailability() {
+		return errors.New("Zipkin Port(9411) is not-setup. failing Zipkin Trace retrieval")
+	}
+
+	sc.logger.Infof("Logging Zipkin Trace: %s", traceID)
+
+	zipkinTraceEndpoint := fmt.Sprintf(zipkin.ZipkinTraceEndpoint, traceID)
+
+	// Sleep to ensure all traces are correctly pushed on the backend.
+	time.Sleep(5 * time.Second)
```

Comment:
We had a discussion about the danger of sleeping here. What are the alternatives?

Reply:
Discussion in the design doc was around requiring sleep for the client side spans to be pushed to zipkin. However we are no longer pushing client side spans(doc was updated) as only adding the trace header suffices to get details about a request. While testing I discovered that there is a lag even for back-end components(esp istio-policy) where spans were not pushed for few seconds. A 5 second sleep guaranteed that we see all the details for a request hence it is added here in the LogZipkinTrace(..) method. Since these are backend component we don't have any control over them once the request is complete.

Reply:
I would challenge the "guaranteed" statement. Let me insist one more time: is it really impossible to  check (for sure) that all traces were pushed?

Reply:
So the process for trace propagation is:

1. On start of trace the trace ID is posted to Zipkin backend without any spans associated with the trace.
2. After the request is completed, as an out-of-band operation, zipkin instrumentations pushes the span information to Zipkin backend.

Since the span propagation is out-of-band and this is handled by the Zipkin instrumentation library, client doesn't have access to it nor there is an API call to force it. (Zipkin API doc: https://zipkin.io/zipkin-api/#/) As the traceId itself is pushed in Step 1, if a GET call is made on that traceId to Zipkin backend, the call itself will not fail but returns only spans for that trace that have been pushed by the participating components.

I think I loosely used the word "guaranteed" it should be more "Safe to say" that spans will be propagated. We are also logging the TraceID in case if the spans are not pushed in 5 secs.

Reply:
Is 5 sec the recommended timeout period? Maybe we can create a polling operation that does the GET call till a timeout for every second or so, instead?

Reply:
We need to make sure (2) happens before shutting down the test environment. Waiting and hoping that it's done should be the last resort. This sounds like a too common task to me, so that we're the only ones affected by the lack of a deterministic behavior. If the API doesn't provide the feature, maybe there's a signal you can send, or a graceful shutdown that ensures the completion? As a last resort, there's Srinivas' suggestion of polling the server to ensure the operations were completed.

Reply:
I don't think poll will help here, as GET even with no spans would not fail because of Step 1.  Having said that I will dig further and also reach out to the monitoring team.

Reply:
After investigating further and reaching out to OpenZipkin community forum ((https://gitter.im/openzipkin/zipkin?at=5bacc7b1cdc50131725f4abd) and talking to monitoring team, it was identified that currently, there is no way to force server side components to publish spans. This mainly stems from the fact that Zipkin being distributed there is no centralized authority that can enforce this behavior. Due to this limitation, waiting by sleep to allow components to push their spans is the best we can currently do.